### ISSUE TITLE
Unvanquished failed to start on Slackware 14.2

### ISSUE THREAD

**Issue by ghost** [Unvanquished/Unvanquished#944]
```
Unvanquished 0.50 Linux x86_64 Jul  2 2016
cmdline: -libpath /usr/lib64/unvanquished -pakpath /usr/share/games/unvanquished/pkg
[FS] Lib path: /usr/lib64/unvanquished 
[FS] Home path: /home/dslackw/.unvanquished 
[FS] Pak search path: /usr/share/games/unvanquished/pkg 
[FS] Pak search path: /home/dslackw/.unvanquished/pkg 
[FS] Loading pak '/usr/share/games/unvanquished/pkg/unvanquished_0.50.0.pk3'... 
Warn: [FS] Could not find pak 'unvanquished' with version '0.49.0' required by '/usr/share/games/unvanquished/pkg/unvanquished_0.50.0.pk3' 
Failed to load pak 'unvanquished': Missing dependency

Could not load base pak 'unvanquished', falling back to default

couldn't exec 'default.cfg' 
Warn: Couldn't read conhistory: No such file or directory 
----- Client Initialization ----- 
Loading RSA keys from pubkey 
Daemon RSA public-key found. 
----- Client Initialization Complete ----- 
Calling GetRefAPI… 
SDL_Init( SDL_INIT_VIDEO )...  
Using SDL Version 2.0.4 
]ATTENTION: default value of option force_s3tc_enable overridden by environment.
SDL using driver "x11" 
Initializing OpenGL display 
Display aspect: 1.778 
...setting mode -2: 
 1920 1080 
The change will take effect after restart. 
The change will take effect after restart. 
Warn: GLEW version < 2.0.0 doesn't support GL core profiles 
Using 24 Color bits, 24 depth, 8 stencil display. 
Using GLEW 1.13.0 
Using enhanced (GL3) Renderer in GL 2.x mode... 
Available modes: '1920x1080 ' 
GL_RENDERER: Gallium 0.4 on AMD RV730 (DRM 2.43.0, LLVM 3.8.0) 
Detected graphics driver class 'Mesa' 
Detected graphics hardware class 'AMD Radeon DX10-class' 
Initializing OpenGL extensions 
...ignoring GL_ARB_debug_output 
...found shading language version 130 
...using GL_ARB_half_float_pixel 
...using GL_ARB_texture_float 
...using GL_EXT_texture_integer 
...using GL_ARB_texture_rg 
...using GL_EXT_texture_filter_anisotropic 
...using GL_ARB_half_float_vertex 
...using GL_ARB_framebuffer_object 
...GL_ARB_get_program_binary found, but with no binary formats 
...using GL_ARB_buffer_storage 
...using GL_ARB_uniform_buffer_object 
...using GL_ARB_map_buffer_range 
...using GL_ARB_sync 
Warn: SSAO not used because GL_ARB_texture_gather is not available. 
glsl shaders took 8744 msec to build 
Warn: R_InitImages: could not load 'gfx/2d/consolechars' 
[FS] Loading pak '/usr/share/games/unvanquished/pkg/unvanquished_0.50.0.pk3'... 
Warn: [FS] Could not find pak 'unvanquished' with version '0.49.0' required by '/usr/share/games/unvanquished/pkg/unvanquished_0.50.0.pk3' 
Failed to load pak 'unvanquished': Missing dependency

Could not load base pak 'unvanquished', falling back to default
```

https://postimg.org/image/46alnaje9/

**Comment by Viech** [comment]
We do not distirbute the Slackware package, so we cannot fix this (as @illwieckz pointed out).

**Comment by dslackw** [comment]
Hi again,
I am a maintainer for Slackware package.
After download all data package in $HOME directory i got this message:
https://postimg.org/image/prbpy8hrr/

**Comment by Viech** [comment]
Hello @dslackw that appears to be a different error, I'm not sure if this can also happen when the data packages (including the gamelogic) don't match the engine binaries (@Kangz, @Amanieu?).

Can we get the console output in text form so that we can see which packages actually get loaded?

**Comment by dslackw** [comment]
Here it is:
http://pastebin.com/wGjPzya0

**Comment by Viech** [comment]
It does seem like all relevant files are present. I noticed that it uses `cgame-x86_64.nexe`. The two developers I just pinged should be able to tell if this a bug worthy a new issue or just a common configuration issue.

**Comment by dslackw** [comment]
Does this issue should be reopened?

**Comment by Viech** [comment]
Yeah, I guess we can re-open the issue as the packaging looks good to me now and hence it may be a bug on our end.

**Comment by dslackw** [comment]
Yes the package builds normally. So I will wait for a solution from here.
Thank you

**Comment by Viech** [comment]
It may be that the virtual machine crashes. Can you try
- launching `daemon` with `-set vm.cgame.debugLoader 5`, which will create `cgame-nacl_loader.log` in Unvanquished's folder in your home directory with more details on the crash, potentially,
- launching `daemon` with `-set vm.cgame.type 2`, given that the `cgame-native-exe` (or so) that you compiled is in the same directory as `daemon`?

**Comment by dslackw** [comment]
Creates only the file ```cgame-nacl_loader.log``` without contents.
https://postimg.org/image/t1117mlep/

**Comment by Kangz** [comment]
How about vm.cgame.type 2? If that crashes too, can you give us a backtrace?

**Comment by Amanieu** [comment]
@dslackw you need to run with vm.cgame.type 0 (the default) for vm.cgame.debugLoader to work

**Comment by dslackw** [comment]
I try but without result:
http://pastebin.com/CraPG3vS
https://postimg.org/image/9y12jnptj/

**Comment by dslackw** [comment]
Is there any developments in this issue?

**Comment by Viech** [comment]
The error you get is very generic and unexplaining. As a package maintainer you will have a rather specific setup that may be hard for us to reproduce exactly. Ideally you would provide us with a backtrace of the crash. However, I'm no expert on the plenty of options in which you can start a gamelogc and on how to debug a specific one among those. I can just tell that I personally run the game within a debugger (e.g. `gdb`) with `-set vm.cgame.type 3 -set vm.cgame.debug 1` if I want to debug the module that's crashing for you. However, it is then loaded from a shared object file (the `cgame-native-dll.so` that you may or may not have compiled, see `ccmake` and your build directory), which may or may not crash for you. So far we only know that the platform independent VM file shipped inside the pk3 crashes for you. It may be that you can compile and use breakpad support to get a binary crash dump in this case. @DolceTriade and @slipher should elaborate on that option in case you don't get a crash registered within `gdb`.

**Comment by slipher** [comment]
If there is a crash in the NaCl cgame, it already should have created crash dumps under `~/.unvanquished/crashdump`. A crash in the engine or a shared object can be dumped if Breakpad is enabled by setting the `USE_BREAKPAD` option with CMake (but you could simply use gdb for such cases). Note that by a crash I mean a kind of fault that would trigger a signal handler.

**Comment by Incognito4nonymous** [comment]
Hello, Slackware 14.2 user here, I installed the unvanquished package and when I run it normally i get this error : 
Warn: Built-in shaders failed: Couldn't compile vertex shader: deformVertexes
And when I run it with my Nvidia GPU the error is : 
Warn: Built-in shaders failed: Couldn't compile fragment shader: depthtile1
I used both versions, the prebuilt and the SlackBuild compiled one.
I also tried on another computer using Slackware 14.1 and on Windows, nothing works.
By the way aren't you the guy who made that Fluxbox+Slackware Video on youtube ? :smiley:

**Comment by illwieckz** [comment]
Can you try the [prebuilt binary](https://github.com/Unvanquished/Unvanquished/releases) and try to reproduce the issue with it to figure out if it's a building or packaging issue on slackware side or a bug in unvanquished?

**Comment by illwieckz** [comment]
@Incognito4nonymous there is nothing in this thread that proves the slackware package your are trying to use is already working for someone else. It's a third-party effort we can't verify. @dslackw is your package works on your side?

**Comment by Incognito4nonymous** [comment]
Oh blast, that's a big file.
I'll ask a friend to download it, and I'll come back to give my feedback.

**Comment by illwieckz** [comment]
You can download it with bittorrent using that [torrent file](https://cdn.unvanquished.net/latest.php). This way you just disable all the `map-*.pk3` and the `tex-*.pk3` downloads. Of course you will not be able to load a map to play but you would be able to load the game itself. You can of course disable the `win32.zip`, `win64.zip` and `mac.zip` downloads too.

**Comment by illwieckz** [comment]
If you download `map-vega_b4.pk3`, `tex-common_2.0.pk3`, and `tex-vega_b4.pk3`, you would be able to load the _vega_ map and do a complete unvanquished experience without downloading everything else. This way you will save 312M of download (you can fetch them later).

**Comment by dslackw** [comment]
@Incognito4nonymous yes i'am 
@illwieckz nope, release 0.50.0 build perfect but not run :(

**Comment by Incognito4nonymous** [comment]
Nice to meet you ! Your video made me dicover trance !
And actually the issue ain't Slackware specific, it also affects Windows, as I mentioned above I downloaded the multiplatform version first, and afterwards I used the Slackbuild since it worked neither on Linux than on Windows.
All the pkgs where downloaded from here.
So I think it is another issue. The Slackbuild builds flawlessly without any errors.

**Comment by Incognito4nonymous** [comment]
So... No one has any idea of how I can solve this ? As I said I already downloaded the GitHub version before downloading the SlackBuild, so the issue ain't OS specific.
I really wanna play that game hard as I plan to use it's engine for some projects.

**Comment by dslackw** [comment]
I remember playing this game in the past... it is really awesome !!!

**Comment by Incognito4nonymous** [comment]
Can you try with the moddb version ? The actual one seems to be broken.

**Comment by dslackw** [comment]
I downloaded version 42 from here, it is right ?
The same error.

**Comment by Viech** [comment]
The release on ModDB is outdated, the current version is Alpha 50.

Also @Incognito4nonymous please don't hijack other issues. Your error message is completely different from what @dslackw is experiencing. This issue is only for tracking @dslackw's problem. To clean this up, here are the relevant logs again:

http://pastebin.com/CraPG3vS
https://postimg.org/image/9y12jnptj/

As I said before, the error message is a rather generic one telling us that some virtual machine crashed. We need a backtrace as discussed before to find the underlying problem.

**Comment by Incognito4nonymous** [comment]
Yeah, my confusion came out of the fact that I'm also using Slackware 14.2.

**Comment by dslackw** [comment]
Hi again,
After a successful building the new version 0.51.1 on Slackware 14.2:
https://pastebin.com/iV9anzfT

It still failed to start the game
Image: https://ibb.co/Lhw7GbR
Pastebin: https://pastebin.com/EwNY11Rv

Slackbuild: https://gitlab.com/dslackw/SlackBuilds/tree/master/unvanquished

Any advice?
Thanks

**Comment by illwieckz** [comment]
This can be related to missing or miconfigured `/dev/shm`. On my distro it's mounted like this:

```
$ mount | grep shm
tmpfs on /dev/shm type tmpfs (rw,nosuid,nodev)
```

This is related to NativeClient I think.

**Comment by dslackw** [comment]
In my own:
$ mount | grep shm                                                     
tmpfs on /dev/shm type tmpfs (rw)

**Comment by illwieckz** [comment]
Svoboda#1890 @ Discord is facing the same problem on Slackware.
Running `strace -f ~/.local/share/unvanquished/base/daemon` disclosed this:

```
[pid 25565] execve("/home/svoboda/.local/share/unvanquished/base/nacl_helper_bootstrap", ["/home/svoboda/.local/share/unvan"..., "/home/svoboda/.local/share/unvan"..., "--r_debug=0xXXXXXXXXXXXXXXXX", "--reserved_at_zero=0xXXXXXXXXXXX"..., "-B", "/home/svoboda/.local/share/unvan"..., "-e", "-i", "100:28", "--", "/home/svoboda/.local/share/unvan"..., "100"], 0x7ffe2b3e2cd0 /* 0 vars */ <unfinished ...>
[pid 25530] <... vfork resumed>)        = 25565
[pid 25530] close(30)                   = 0
[pid 25565] <... execve resumed>)       = -1 EPERM (Operation not permitted)
[pid 25530] read(29,  <unfinished ...>
[pid 25565] --- SIGSEGV {si_signo=SIGSEGV, si_code=SI_KERNEL, si_addr=NULL} ---
[pid 25530] <... read resumed>"", 4)    = 0
[pid 25530] close(29 <unfinished ...>
[pid 25565] +++ killed by SIGSEGV +++
```

**Comment by slipher** [comment]
I reproduced the bug in a Slackware 14.2 VM. There seems to be some sort of kernel bug where the `execve` system call fails, and delivers SIGSEGV, if the target is the `nacl_helper_bootstrap` binary and the user is non-root. It can most directly be seen by running `./nacl_helper_bootstrap`, which prints a usage string if the user is root, and `Segmentation fault` otherwise.

This possibility is mentioned by the [execve man page](https://man7.org/linux/man-pages/man2/execve.2.html):

> In most cases where execve() fails, control returns to the original
       executable image, and the caller of execve() can then handle the
       error.  However, in (rare) cases (typically caused by resource
       exhaustion), failure may occur past the point of no return: the
       original executable image has been torn down, but the new image could
       not be completely built.  In such cases, the kernel kills the process
       with a SIGSEGV (SIGKILL until Linux 3.17) signal.

strace and gdb didn't work on this binary, so I had to use a custom ptrace program:

```c
#include <sys/ptrace.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>
#include <sys/reg.h>
#include <sys/user.h>
#include <stdio.h>
#include <sys/syscall.h>
#include <errno.h>

int main(int argc, char** argv)
{
    if (argc < 2) return 1;
    pid_t child = fork();
    if(child == 0) {
        if (0 != ptrace(PTRACE_TRACEME, 0, NULL, NULL)) {
            write(1, "fail\n", 5);
            perror("traceme");
            return 2;
        }
        raise(SIGSTOP);
        execl(argv[1], argv[1], NULL);
        char buf[99];
        int l = sprintf(buf, "ouch %d\n", errno);
        write(1, buf, l);
    }
    else {
        for (;;) {
            int status;
            if (waitpid(child, &status, 0) != child) {
                perror("waitpid");
                break;
            }
            if (!WIFSTOPPED(status)) break;
            printf("stopsig %d\n", WSTOPSIG(status));
            if (WSTOPSIG(status) != SIGSTOP && WSTOPSIG(status) != SIGTRAP)
                break;
            struct user_regs_struct regs; 
            if (0 != ptrace(PTRACE_GETREGS, child, NULL, &regs)) {
                perror("ptrace regs");
                break;
            }
            printf("system call %ld rax=%ld from pid %d\n", regs.orig_rax, regs.rax, child);
            if (0 != ptrace(PTRACE_SYSCALL, child, NULL, NULL)) {
                perror("ptrace_syscall");
                break;
            }
        }
    }
    return 0;
}
```

It gives this output when run as a non-root user:

```
bash-4.3$ ./ptracedemo ../unvlinux51/nacl_helper_bootstrap     
stopsig 19
system call 234 rax=0 from pid 5829
stopsig 5
system call 59 rax=-38 from pid 5829
stopsig 5
system call 59 rax=-1 from pid 5829
stopsig 11
```

Compare to when run as root:

```
bash-4.3$ sudo ./ptracedemo ../unvlinux51/nacl_helper_bootstrap
stopsig 19
system call 234 rax=0 from pid 5823
stopsig 5
system call 59 rax=-38 from pid 5823
stopsig 5
system call 59 rax=0 from pid 5823
stopsig 5
system call 59 rax=0 from pid 5823
stopsig 5
system call 20 rax=-38 from pid 5823
bootstrap_helper: Usage: PROGRAM ARGS...
stopsig 5
system call 20 rax=41 from pid 5823
stopsig 5
system call 231 rax=-38 from pid 5823
```

**Comment by DolceTriade** [comment]
It's less likely to be a kernel bug than a sysctl somewhere that slackware probably has enabled by default.

**Comment by slipher** [comment]
Thanks to @DolceTriade's hint, I found the way to fix the executable loading problem: `sudo /sbin/sysctl -w vm.vmmap_min_addr=65536`. (Not sure if something else is needed to persist this across reboot.) This enables `nacl_helper_bootstrap` to execute - it requests its code section to be loaded at 0x10000 which is smaller than the default minimum address of 98304.

I still faced the issue of a nonexistent shm filesystem on my machine, but hopefully others who have this part in order would be able to start Unvanquished.

**Comment by illwieckz** [comment]
> I found the way to fix the executable loading problem: `sudo /sbin/sysctl -w vm.vmmap_min_addr=65536`. (Not sure if something else is needed to persist this across reboot.)

This will not survive next reboot indeed, here is the persistent way (as root):

```sh
echo 'vm.vmmap_min_addr=65536' > '/etc/sysctl.d/30-unvanquished-nacl-vmmap.conf'
sysctl -p '/etc/sysctl.d/90-unvanquished-nacl-vmmap.conf'
```

First command is to store the option in a way it will be automatically set on next reboot, second command is to read the option right now using the file expected to be read on next reboot (makes easier to check this file does not have syntax error and work before rebooting).

The file name to be stored in `/etc/sysctl.d` can be named the way you want but may have to end with `.conf` extension in all cases.

**Comment by slipher** [comment]
https://github.com/DaemonEngine/Daemon/pull/406 proposes to add code which checks the sysctl value, and if it is too large, aborts the program and suggests a fix along the lines of @illwieckz's previous comment.

**Comment by illwieckz** [comment]
This is assumed to be fixed by DaemonEngine/Daemon#406
